File: src/evidence/independence.ts
/** Profile-selected evidence lineage and independence policy (FR-047). */

import { readFileSync } from "node:fs";

import { z } from "zod";

import type {
  Binding,
  EvidenceLineage,
  IndependenceAssessment,
  IndependenceDimension,
  IndependencePolicy,
  IndependenceRequirement,
} from "./types.js";

const IndependenceDimensionSchema = z.enum([
  "actor",
  "implementation-toolchain",
  "technique",
  "data-source",
  "review-path",
]);

export const EvidenceLineageSchema = z
  .object({
    actor: z.string().trim().min(1).optional(),
    implementationToolchain: z.string().trim().min(1).optional(),
    technique: z.string().trim().min(1).optional(),
    dataSource: z.string().trim().min(1).optional(),
    reviewPath: z.string().trim().min(1).optional(),
  })
  .strict()
  .refine((lineage) => Object.keys(lineage).length > 0, {
    message: "must name at least one lineage dimension",
  });

const IndependenceRequirementSchema = z
  .object({
    id: z.string().trim().min(1),
    obligation: z.string().trim().min(1),
    dimensions: z.array(IndependenceDimensionSchema).min(1),
    rationale: z.string().trim().min(1),
  })
  .strict()
  .superRefine((requirement, context) => {
    if (
      new Set(requirement.dimensions).size !== requirement.dimensions.length
    ) {
      context.addIssue({
        code: "custom",
        path: ["dimensions"],
        message: "contains duplicate dimensions",
      });
    }
  });

export const IndependencePolicySchema = z
  .object({
    schemaVersion: z.literal(1),
    profile: z
      .string()
      .trim()
      .regex(/^AP-[0-9]+$/),
    requirements: z.array(IndependenceRequirementSchema).min(1),
  })
  .strict()
  .superRefine((policy, context) => {
    const ids = new Set<string>();
    const obligations = new Set<string>();
    for (const [index, requirement] of policy.requirements.entries()) {
      if (ids.has(requirement.id)) {
        context.addIssue({
          code: "custom",
          path: ["requirements", index, "id"],
          message: "duplicates another requirement id",
        });
      }
      if (obligations.has(requirement.obligation)) {
        context.addIssue({
          code: "custom",
          path: ["requirements", index, "obligation"],
          message: "duplicates another obligation requirement",
        });
      }
      ids.add(requirement.id);
      obligations.add(requirement.obligation);
    }
  });

export function validateEvidenceLineage(value: unknown): EvidenceLineage {
  return parse(EvidenceLineageSchema, value, "evidence lineage");
}

export function validateIndependencePolicy(value: unknown): IndependencePolicy {
  return parse(IndependencePolicySchema, value, "independence policy");
}

export function readEvidenceLineage(path: string): EvidenceLineage {
  return validateEvidenceLineage(readJson(path, "evidence lineage"));
}

export function readIndependencePolicy(path: string): IndependencePolicy {
  return validateIndependencePolicy(readJson(path, "independence policy"));
}

/** Refuse a stale/typoed projection rather than silently evaluating nothing. */
export function requireKnownPolicyObligations(
  policy: IndependencePolicy,
  knownObligations: Iterable<string>,
): void {
  const known = new Set(knownObligations);
  const unknown = policy.requirements
    .map((requirement) => requirement.obligation)
    .filter((obligation) => !known.has(obligation))
    .sort(compare);
  if (unknown.length > 0) {
    throw new Error(
      `independence policy names unknown obligation(s): ${unknown.join(", ")}`,
    );
  }
}

/**
 * Find two distinct evidence relationships separated on every selected axis.
 *
 * A dimension absent on either side cannot demonstrate separation. Two suites
 * sharing one model/parser/corpus therefore remain visibly common-mode when the
 * profile selects the corresponding dimension.
 */
export function assessIndependence(
  profile: string,
  requirement: IndependenceRequirement,
  bindings: Binding[],
): IndependenceAssessment {
  const ordered = [...bindings].sort((a, b) => compare(a.suite, b.suite));
  const dimensions = requirement.dimensions.map((dimension) => {
    const values = new Set<string>();
    const missingSuites: string[] = [];
    for (const binding of ordered) {
      const value = valueFor(binding.lineage, dimension);
      if (value === undefined) missingSuites.push(binding.suite);
      else values.add(value);
    }
    return {
      dimension,
      values: [...values].sort(compare),
      missingSuites,
    };
  });

  for (let left = 0; left < ordered.length; left += 1) {
    for (let right = left + 1; right < ordered.length; right += 1) {
      const a = ordered[left];
      const b = ordered[right];
      if (
        a.suite !== b.suite &&
        requirement.dimensions.every((dimension) => {
          const first = valueFor(a.lineage, dimension);
          const second = valueFor(b.lineage, dimension);
          return (
            first !== undefined && second !== undefined && first !== second
          );
        })
      ) {
        return {
          profile,
          requirement: requirement.id,
          obligation: requirement.obligation,
          status: "satisfied",
          dimensions,
          satisfiedBy: [a.suite, b.suite],
          summary:
            `${requirement.id} is satisfied by ${a.suite} and ${b.suite}; ` +
            `they differ on ${requirement.dimensions.join(", ")}.`,
        };
      }
    }
  }

  const detail = dimensions
    .map((item) => {
      const missing =
        item.missingSuites.length === 0
          ? ""
          : `; missing on ${item.missingSuites.join(", ")}`;
      return `${item.dimension}: ${item.values.length} distinct${missing}`;
    })
    .join("; ");
  return {
    profile,
    requirement: requirement.id,
    obligation: requirement.obligation,
    status: "insufficient",
    dimensions,
    summary:
      `${requirement.id} requires two evidence lines differing on every selected ` +
      `dimension (${requirement.dimensions.join(", ")}); ${detail}.`,
  };
}

function valueFor(
  lineage: EvidenceLineage | undefined,
  dimension: IndependenceDimension,
): string | undefined {
  let value: unknown;
  switch (dimension) {
    case "actor":
      value = lineage?.actor;
      break;
    case "implementation-toolchain":
      value = lineage?.implementationToolchain;
      break;
    case "technique":
      value = lineage?.technique;
      break;
    case "data-source":
      value = lineage?.dataSource;
      break;
    case "review-path":
      value = lineage?.reviewPath;
      break;
  }
  if (typeof value !== "string" || value.trim().length === 0) return undefined;
  return value.trim();
}

function readJson(path: string, label: string): unknown {
  try {
    return JSON.parse(readFileSync(path, "utf8"));
  } catch (cause) {
    throw new Error(
      `${label} at ${path} is not readable JSON: ` +
        (cause instanceof Error ? cause.message : String(cause)),
    );
  }
}

function parse<T>(schema: z.ZodType<T>, value: unknown, label: string): T {
  const parsed = schema.safeParse(value);
  if (parsed.success) return parsed.data;
  const detail = parsed.error.issues
    .map((issue) => `${issue.path.join(".") || "record"}: ${issue.message}`)
    .join("; ");
  throw new Error(`invalid ${label}: ${detail}`);
}

function compare(a: string, b: string): number {
  return a === b ? 0 : a < b ? -1 : 1;
}

File: rust/crates/quoin-evidence/src/independence.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Agent-IX

//! Profile-selected evidence lineage and independence policy (FR-094).
//!
//! The question is always "do two evidence lines exist that differ on *every*
//! axis this profile selected". A dimension absent on either side cannot
//! demonstrate separation, so two suites sharing one model, parser or corpus
//! stay visibly common-mode rather than passing by omission.

use std::collections::BTreeSet;

use crate::error::EvidenceError;
use crate::ids::{ObligationId, ProfileId, SuiteId};
use crate::types::{
    Binding, EvidenceLineage, IndependenceAssessment, IndependenceDimension,
    IndependenceDimensionAssessment, IndependencePolicy, IndependenceRequirement,
    IndependenceStatus,
};

/// Check one policy against the FR-094 boundary.
///
/// # Errors
///
/// [`EvidenceError::InvalidRecord`] carrying every clause that failed, joined
/// with `; `.
pub fn validate_independence_policy(policy: &IndependencePolicy) -> Result<(), EvidenceError> {
    let mut clauses: Vec<String> = Vec::new();
    if policy.schema_version != 1 {
        clauses.push("schemaVersion: must be 1".to_owned());
    }
    if policy.requirements.is_empty() {
        clauses.push("requirements: must not be empty".to_owned());
    }
    let mut ids: BTreeSet<&str> = BTreeSet::new();
    let mut obligations: BTreeSet<&str> = BTreeSet::new();
    for (index, requirement) in policy.requirements.iter().enumerate() {
        if requirement.id.trim().is_empty() {
            clauses.push(format!("requirements.{index}.id: must not be empty"));
        }
        if requirement.obligation.as_str().trim().is_empty() {
            clauses.push(format!(
                "requirements.{index}.obligation: must not be empty"
            ));
        }
        if requirement.rationale.trim().is_empty() {
            clauses.push(format!("requirements.{index}.rationale: must not be empty"));
        }
        if requirement.dimensions.is_empty() {
            clauses.push(format!(
                "requirements.{index}.dimensions: must not be empty"
            ));
        }
        let unique: BTreeSet<IndependenceDimension> =
            requirement.dimensions.iter().copied().collect();
        if unique.len() != requirement.dimensions.len() {
            clauses.push(format!(
                "requirements.{index}.dimensions: contains duplicate dimensions"
            ));
        }
        if !ids.insert(requirement.id.as_str()) {
            clauses.push(format!(
                "requirements.{index}.id: duplicates another requirement id"
            ));
        }
        if !obligations.insert(requirement.obligation.as_str()) {
            clauses.push(format!(
                "requirements.{index}.obligation: duplicates another obligation requirement"
            ));
        }
    }
    if clauses.is_empty() {
        Ok(())
    } else {
        Err(EvidenceError::invalid(
            "independence policy",
            clauses.join("; "),
        ))
    }
}

/// Check one lineage against the FR-094 boundary.
///
/// # Errors
///
/// [`EvidenceError::InvalidRecord`] when the lineage names nothing at all: an
/// empty object claims to have recorded separation facts while recording none.
pub fn validate_evidence_lineage(lineage: &EvidenceLineage) -> Result<(), EvidenceError> {
    if lineage.is_empty() {
        return Err(EvidenceError::invalid(
            "evidence lineage",
            "record: must name at least one lineage dimension",
        ));
    }
    Ok(())
}

/// Refuse a stale or typoed projection rather than silently evaluating nothing.
///
/// A policy whose obligation was renamed would otherwise report every
/// requirement as vacuously assessed, which is the opposite of what an
/// independence check is for.
///
/// # Errors
///
/// [`EvidenceError::PolicyUnknownObligations`], naming the unknown ids sorted.
pub fn require_known_policy_obligations<'a, I>(
    policy: &IndependencePolicy,
    known: I,
) -> Result<(), EvidenceError>
where
    I: IntoIterator<Item = &'a str>,
{
    let known: BTreeSet<&str> = known.into_iter().collect();
    let unknown: BTreeSet<&str> = policy
        .requirements
        .iter()
        .map(|requirement| requirement.obligation.as_str())
        .filter(|obligation| !known.contains(obligation))
        .collect();
    if unknown.is_empty() {
        return Ok(());
    }
    Err(EvidenceError::PolicyUnknownObligations {
        obligations: unknown.into_iter().collect::<Vec<_>>().join(", "),
    })
}

/// Find two distinct evidence relationships separated on every selected axis.
///
/// The bindings are ordered by suite first, so the pair reported for a given
/// binding set does not depend on the order the caller assembled them in.
#[must_use]
pub fn assess_independence(
    profile: &ProfileId,
    requirement: &IndependenceRequirement,
    bindings: &[Binding],
) -> IndependenceAssessment {
    let mut ordered = bindings.to_vec();
    ordered.sort_by(|a, b| a.suite.as_str().cmp(b.suite.as_str()));

    let dimensions: Vec<IndependenceDimensionAssessment> = requirement
        .dimensions
        .iter()
        .map(|dimension| {
            let mut values: BTreeSet<String> = BTreeSet::new();
            let mut missing = Vec::new();
            for binding in &ordered {
                match value_for(binding.lineage.as_ref(), *dimension) {
                    Some(value) => {
                        values.insert(value);
                    }
                    None => missing.push(binding.suite.clone()),
                }
            }
            IndependenceDimensionAssessment {
                dimension: *dimension,
                values: values.into_iter().collect(),
                missing_suites: missing,
            }
        })
        .collect();

    let selected = requirement
        .dimensions
        .iter()
        .map(|dimension| dimension.as_str())
        .collect::<Vec<_>>()
        .join(", ");

    for (left, a) in ordered.iter().enumerate() {
        for b in ordered.iter().skip(left + 1) {
            if a.suite == b.suite {
                continue;
            }
            let separated = requirement.dimensions.iter().all(|dimension| {
                match (
                    value_for(a.lineage.as_ref(), *dimension),
                    value_for(b.lineage.as_ref(), *dimension),
                ) {
                    (Some(first), Some(second)) => first != second,
                    _ => false,
                }
            });
            if !separated {
                continue;
            }
            return IndependenceAssessment {
                profile: profile.clone(),
                requirement: requirement.id.clone(),
                obligation: requirement.obligation.clone(),
                status: IndependenceStatus::Satisfied,
                dimensions,
                satisfied_by: Some((a.suite.clone(), b.suite.clone())),
                summary: format!(
                    "{} is satisfied by {} and {}; they differ on {selected}.",
                    requirement.id, a.suite, b.suite
                ),
            };
        }
    }

    let detail = dimensions
        .iter()
        .map(|item| {
            let missing = if item.missing_suites.is_empty() {
                String::new()
            } else {
                format!(
                    "; missing on {}",
                    item.missing_suites
                        .iter()
                        .map(SuiteId::as_str)
                        .collect::<Vec<_>>()
                        .join(", ")
                )
            };
            format!(
                "{}: {} distinct{missing}",
                item.dimension,
                item.values.len()
            )
        })
        .collect::<Vec<_>>()
        .join("; ");

    IndependenceAssessment {
        profile: profile.clone(),
        requirement: requirement.id.clone(),
        obligation: requirement.obligation.clone(),
        status: IndependenceStatus::Insufficient,
        dimensions,
        satisfied_by: None,
        summary: format!(
            "{} requires two evidence lines differing on every selected dimension ({selected}); {detail}.",
            requirement.id
        ),
    }
}

/// The recorded value for one dimension, trimmed, with blank treated as absent.
///
/// A whitespace-only lineage entry is not a separation fact, and treating it as
/// one would let two suites "differ" on a pair of empty strings.
fn value_for(
    lineage: Option<&EvidenceLineage>,
    dimension: IndependenceDimension,
) -> Option<String> {
    let lineage = lineage?;
    let raw = match dimension {
        IndependenceDimension::Actor => lineage.actor.as_deref(),
        IndependenceDimension::ImplementationToolchain => {
            lineage.implementation_toolchain.as_deref()
        }
        IndependenceDimension::Technique => lineage.technique.as_deref(),
        IndependenceDimension::DataSource => lineage.data_source.as_deref(),
        IndependenceDimension::ReviewPath => lineage.review_path.as_deref(),
    }?;
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        None
    } else {
        Some(trimmed.to_owned())
    }
}

/// The obligation ids a policy names, in policy order.
#[must_use]
pub fn policy_obligations(policy: &IndependencePolicy) -> Vec<ObligationId> {
    policy
        .requirements
        .iter()
        .map(|requirement| requirement.obligation.clone())
        .collect()
}
